### ISSUE TITLE
Is OneDriveClient concurrency-safe?

### ISSUE THREAD

**Issue by tipa** [OneDrive/onedrive-sdk-csharp#163]
Is the OneDriveClient safe to use for up-/downloading multiple objects at once (via multiple tasks) like in the example below? And is there any best-practice for up-/downloadng lots of files (at once/asap)?

```csharp
List<Task> tasks = new List<Task>();
foreach (var itemPath in items)
{
  tasks.Add(Task.Run(async () =>
    {
        using (var itemStream = await oneDriveClient.Drive.Special.AppRoot.ItemWithPath(itemPath)
           .Content.Request().GetAsync())
        {   ...    }
    }));
}
Task.WaitAll(tasks.ToArray());
```

[slight edits by @cmayer for readability]

**Comment by cdmayer** [comment]
Your example is good. Generally speaking, the default `HttpProvider` included in Graph.Core is safe for multiple requests at once. [You can inspect the source here](https://github.com/microsoftgraph/msgraph-sdk-dotnet/blob/master/src/Microsoft.Graph.Core/Requests/HttpProvider.cs). Let me know if you run into any specific issues.

**Comment by cdmayer** [comment]
One other comment, when you construct the `OneDriveClient` object, be sure to explicitly pass in an instance of `HttpProvider`. When sending many requests, this will skip creating and disposing many instances of the class needlessly.

**Comment by tipa** [comment]
Could you please point me to the source code where there is the `HttpProvider` created repeatedly if none is passed in to the `OneDriveClient`? I searched the source but wasn't successful finding it.

**Comment by cdmayer** [comment]
Ah, I was mistaken. In [BaseClient](https://github.com/microsoftgraph/msgraph-sdk-dotnet/blob/master/src/Microsoft.Graph.Core/Requests/BaseClient.cs) the `HttpProvider` is created in the default constructor. I had a different part of the code in mind (authentication). Sorry for the mistake!

TL;DR: You can ignore the comment above. It should be safe to pass `null` to the `OneDriveClient` constructor if you want the default `HttpProvider` implementation.
